### PR TITLE
Add an optional `target` argument to ps()

### PR DESCRIPTION
**Pull request by pygy**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ Patchinko exposes 3 functions: `patch`, `scope`, & `ps`.
 
 If any target properties are instances of `scope(function)`, it will supply the scoped function with the target property for that key, and assign the result back to the target.
 
-`ps` is a composition of `patch` & `scope`, for when you need to patch recursively.
+`ps([ target, ] input)` is a composition of `patch` & `scope`, for when you need to patch recursively. If you supply a `target`, the original value will be left untouched (useful for immutable patching).
```
Comment (barneycarroll) [inline README.md]: I was misreading the signature last night and interpreted `[ ...targets,] input`, so I couldn't wrap my head around the desired semantics. Got it now! 👍

File: README.md [hunk 2]
```diff
@@ -68,7 +69,11 @@ p(thing, {
       structure: s(structure =>
         structure.concat('roflmao') // Why not
       )
-    })
+    }),
+    with: ps(
+      [],
+      {1: 'copy'}
+    ) // ['a', 'copy', 'tacked'], the original array is left untouched
```
Comment (barneycarroll) [inline README.md]: The fact this incidentally demonstrates you don't need `scope` to extend arrays detracts from the feature a bit, especially seeing as this all-in-one example is mutative at the top level, so the fact the original array is preserved isn't easy to intuit on an intent level. The docs already needed a re-write but this brings the necessity closer.

File: README.md [hunk 3]
```diff
@@ -81,6 +86,10 @@ Observe that:
 
 `stupidly.deep.stucture` & `utils.fibonacci` show that any kind of structure can be modified or replaced at any kind of depth: `patch` is geared towards the common case of objects, but `scope` can deal with any type in whatever way necessary. You get closures for free so gnarly patch logic can be isolated at the point where it makes the most sense.
 
+```JS
+
+````
+
```
Comment (barneycarroll) [inline README.md]: ?

File: index.js [hunk 4]
```diff
@@ -20,9 +20,13 @@ function scope(closure){
...
-function ps(input){
-  return new scope(function(definition){
-    return patch(definition, input)
+function ps(target, input){
+  return arguments.length === 2
+  ? new scope(function(definition){
+    return patch(target, definition, input)
+  })
+  : new scope(function(definition){
+    return patch(definition, target) // `target` really is `input` in that case
```
Comment (barneycarroll) [inline index.js]: It took me embarrassingly long to figure out how this worked with such a small patch without breaking anything – I was convinced new functions would be necessary but AFAICT, this simple re-jigging of patch signature accounts for all immutable concerns, right? Incredibly neat if so!

File: README.md (hunk 3 above)
Comment (pygy) [reply README.md]: Oh, sorry :-/

File: index.js (hunk 4 above)
Comment (pygy) [reply index.js]: I think so.

I'd immediately follow with `psa = input => ps([], input)` and `pso = input => ps({}, input)` in client code.

File: README.md (hunk 2 above)
Comment (pygy) [reply README.md]: > The fact this incidentally demonstrates you don't need `scope` to extend arrays

Only when you know the indices in advance. This is mostly useful for tuples.

**Pull request merged**